lib/AST/NameLookup.cpp

```diff
@@ -2170,7 +2171,18 @@ ExtendedNominalRequest::evaluate(Evaluator &evaluator,
   auto nominalTypes
     = resolveTypeDeclsToNominal(evaluator, ctx, referenced, modulesFound,
                                 anyObject);
-  return nominalTypes.empty() ? nullptr : nominalTypes.front();
+  switch (nominalTypes.size()) {
+  case 0: return nullptr;
+  case 1: return nominalTypes[0];
+  default:
+    // If there were several references, we will report an ambiguity error
+    // anyways, so this diagnostic is not applicable.
+    if (referenced.size() == 1)
+      // TODO(Varun): How do we get the source range for the extended type?
```

Comment:
Could you grab it from the extended type repr? (i.e. `ext->getExtendedTypeRepr().getSourceRange()`?)

Reply:
IIUC, that's not guaranteed to succeed. The same function has a check earlier:
```
  // Prefer syntactic information when we have it.
  if (auto typeRepr = ext->getExtendedTypeRepr()) {
```
so it is possible that we do not have the syntactic information at this point in time.

Reply:
Is it the same with the extended nominal type? I suppose you can skip highlighting the location if it’s not available (unless it’s always not available at this stage).

Reply:
You might want to diagnose this in ExtendedTypeRequest::evaluate() instead of doing it here, because there's no reason to emit the warning for extensions in secondary files.

Also now that you're serializing the extended nominal directly, you can fix the ClangImporter's synthesis of ExtensionDecls to also set the extended nominal, and remove the code from ExtendedNominalRequest for handling the case of a null TypeRepr. It won't happen for parsed extensions.

Reply:
> You might want to diagnose this in ExtendedTypeRequest::evaluate() instead of doing it here, because there's no reason to emit the warning for extensions in secondary files.

Do we have an invariant that ExtendedTypeRequest::evaluate() is not called for extensions in secondary files but ExtendedNominalRequest::evaluate() is?

Reply:
> Also now that you're serializing the extended nominal directly, you can fix the ClangImporter's synthesis of ExtensionDecls to also set the extended nominal, and remove the code from ExtendedNominalRequest for handling the case of a null TypeRepr. It won't happen for parsed extensions.

There's an additional case here here in ParseDecl which can lead to a null TypeRepr:

https://github.com/apple/swift/blob/27ded648cf330d32f421c7aadc1ec132f292f556/lib/Parse/ParseDecl.cpp#L3851-L3855

Any suggestions on how we might get rid of it?

Reply:
> Do we have an invariant that ExtendedTypeRequest::evaluate() is not called for extensions in secondary files but ExtendedNominalRequest::evaluate() might be?

It's not an invariant, but we want to minimize requests in secondary files where possible.

> There's an additional case here here in ParseDecl which can lead to a null TypeRepr:

This is the invalid case with on type written (`extension { ... }`). In this case ExtendedNominalRequest should just return null.